### DISCUSSION TITLE
Why do some glyphs have a four-digit hex code while other have five digits?

### DISCUSSION THREAD

**Discussion by stamminator** [ryanoasis/nerd-fonts#1681] [Q&A] [answered]
I'm troubleshooting an issue where the VS Code integrated terminal does unnecessary line breaks when certain glyphs are present in the output. I've noticed that glyphs with four-digit hex codes don't have the issue, while glyphs with five-digit hex codes do.

![vscode terminal glyph line break issue](https://github.com/user-attachments/assets/4a379e9a-cc91-4e59-ac1c-ea240f339961)

I plan to file an issue with vscode, but first, I want to understand how glyphs work better. Why do the hex codes vary in length? 

EDIT: I've also noticed that when I enter the hex code literal into an [Oh My Posh segment template string](https://ohmyposh.dev/docs/configuration/templates) rather than pasting the icon itself into the template string, the 4-digit one is correct while the 5-digit one isn't.

![image](https://github.com/user-attachments/assets/a68e7280-88fa-4820-8157-882beecc00cc)

I know this is neither the VS Code nor the Oh My Posh repo, but these issues seem conceptually similar.

**Comment by Finii** [comment, marked as answer]
"In the beginning" there was ASCII, where each character can be expressed as 7 bit number - later expanded to 8 bit.
These 8 bits can be expressed as 2 digit hexadecimal number.
But 256 characters is a bit low, considering all these strange letters with appendices like ä and Ó.
Then we got codepages. Etc.

That all was too limited and finally unicode came up. Version 2 had 40'000 characters, nowadays we have 150'000.
The 'address' of a certain character or glyph is its codepoint. And you see for Version 2 with 40'000 (decimal) codepoints each can be expressed as 4 digit hexadecimal number (40000 decimal = `9C40` hex). But this is no longer possible (150'000 decimal = `249F0` hex)

We are not free to use any codepoint we like, there are certain ranges that need to be used. And when the useful-for-Nerdfonts 4 digit range filled up we needed to allocate new glyphs a 5 digit number. Well, in principle.
Codepoints are organized in blocks, see for example https://en.wikipedia.org/wiki/Unicode_block
We can only use codepoints in the _Private Use Areas_ or PUA, and you see in the block overview that there is only a very small block in the 4 digit range and two larger blocks in the 5 and 6 digit range.

That much for history. A long time 4 digit code were all what was needed and this lead to some pitfalls.
So for example in many contexts the `\u` is a prefix for a 4 digit unicode. You see that in your example `\uf1442` is the same as `\uf144` and `2` and this is what you see. Some systems have `\U` for 5/6 digit codes, for some systems you must encode the codepoint in UTF16 that you can put as two `\u` strings. The cheat cheat helps you with these because the calculation is not easy:

![image](https://github.com/user-attachments/assets/beae0a44-4170-4a99-92f4-f16868f7ba53)

Here you see LOL, with codepoint `\Uf1214` or `\udb84\ude14` (the second form is the UTF16 variant that you get by clicking on the UTF16 box).

I have no clue what VS Code uses or expects.

For the unnecessary line breaks, I guess the line breaks because the terminal calculates the length wrong? But the lines seem to be not full. On the other hand the break is not after the special glyph but later on. You need to know that unicode has the concept of different width or characters; they can be single, double, or ambiguous width. Maybe here is a problem.
Is the line break independent of concrete full line length? Why is it at the closing parens? Etc.

I hope this explains at least some things and helps your research. There are for more peculiarities, especially regarding Windows and how it handles stuff. I am not sure if the same holds for VS Code on other platforms and you do not specify your platform. Windows specifically was late in joining the other OSes that all supported 5 digit unicode characters and still struggles (I believe, I can be wrong here).

See also
* #365 (We use Plan 5 now)
* #1059
* https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting#how-do-i-use-glyphs-in-my-terminal

**Reply by stamminator** [reply]
Thank you very much for taking the time to write such an informative response. After reading it, I knew enough to know the right questions to ask, and after a lot of web searching and ChatGPT-assisted Q&A, I think the issue has to do with the glyph's codepoint existing beyond the BMP and being represented as a UTF-16 surrogate pair. Based on my testing and research, it looks like the proper handling of non-BMP characters is just a mess in general.

![Animation](https://github.com/user-attachments/assets/2168aa03-f881-4f14-b586-dcdaea7b0279)

> Is the line break independent of concrete full line length? Why is it at the closing parens? Etc.

The number of above-BMP glyphs is equal to the number of characters pushed into the next line in VS Code's terminal. Here is an example that makes it clearer:

![zqrULEiRZx](https://github.com/user-attachments/assets/fddb8eb7-f416-4b68-b5f5-c06ab178e309)

Using the built-in Windows console host (conhost.exe), we see yet another behavior. Above-BMP glyphs consistently take up two columns, but rather than the wanted glyphs and the unwanted gaps being grouped together, conhost displays each gap next to its glyph.

![conhost](https://github.com/user-attachments/assets/9ed7d28f-b2b4-44ea-a42f-9594feb3631e)

It seems like the safest way to make sure everything "just works" is to use only BMP characters (4-digit codepoint) in terminals.

**Reply by Finii** [reply]
Well, `conhost` is 'broken by design'. Back in the past Microsoft decided to go "full 16 bit characters" which can store all the unicodes (in a world of 4 hex digit unicodes). The number of `short`s stored is the number of characters. Note that it is not UTF16 (as one would expect today) but UCS-2, a severe limitation. Nowadays systems typically store UTF8 strings, where things like `U+F1442` are stored in multiple bytes and the number of bytes is NOT the number of characters anymore. That raises some code complications; not only do you need to parse the content of a string to get its character-length, you also can not just cut it off anywhere you like or you possibly have broken UTF8 sequences in the truncated string.
So back in the time the Microsoft design was rather elegant, simplicity at the cost of more memory consumption.

For more information on "conhost" see [1].

The broken Windows implementation of "5 digit unicodes" was one of the major reasons that kept Nerd Fonts from using the PUA-A to fix 365; but in recent years Microsoft works on phasing that out and the new (standard in Windows 11 iirc) `Windows Terminal` does in fact do an excellent job unicode wise and supports all codepoints.

![image](https://github.com/user-attachments/assets/288f5dbd-6095-45c1-939a-38bf66ea995d)

So in principle 'all' non-Microsoft terminals support PUA-A unicode, and MS is working on its terminals (hence `Windows Terminal`), so that PUA-A support should not be a limitation anymore, except in some very rare circumstances. Usually reporting a (5 hex digit) unicode bug in some Microsoft application is acted apon and not wanted. Although some handling is unspecified like the ambiguous width characters.

If your problems are (just) related to conhost et al, that will probably never be fixed because it would require a complete rewrite from Microsoft's side. But more recent applications are expected to work flawlessly.

For the `\udead` versus `\Ubeef` problem... Both have their drawbacks. The lower case `u` takes exactly 4 hex digits, but the upper case `U` takes as many digits as there are, so `\UbeefF` is not beef and F but some `U+BEEFF` character and you have to express such kind of strings in roundabout way :grimacing: 

[0] https://tonsky.me/blog/unicode/ (esp section _What’s "🤦🏼‍♂️".length?_)
[1] https://devblogs.microsoft.com/commandline/windows-command-line-inside-the-windows-console/ (esp section _Handling Unicode_)
[2] https://devblogs.microsoft.com/commandline/windows-command-line-unicode-and-utf-8-output-text-buffer/

**Reply by stamminator** [reply]
Yeah, I won't bother raising an issue about conhost, but I will submit an issue to the VS Code repo this week describing the PUA-A unicode rendering issue.

> the upper case `U` takes as many digits as there are, so `\UbeefF` is not beef and F but some `U+BEEFF` character and you have to express such kind of strings in roundabout way

Which language is it that uses this syntax for hex literals? YAML solves this by making `U` always 8 digits (`\U0000beefF` = 뻯F), which seems to me to have no downsides if you don't count the extra zeros. Anything to avoid having to do UTF-16 surrogate pair math is a win in my book.

**Reply by Finii** [reply]
Hmm, maybe I'm mistaken. I had the `printf` command in mind, but that is exactly as you say. Maybe it was `echo`?

![image](https://github.com/user-attachments/assets/79d6eb88-2261-49df-b2b3-68b96cb8c46d)

_https://www.man7.org/linux/man-pages/man1/printf.1.html_

**Reply by Finii** [reply]
Ah, with `printf` you have the problem with `\x` at least .... _"1 to 2 digits"_ :grimacing:
And you always have and had this with the octal sequences, which drove me crazy in like 1980 ;-)

**Reply by Finii** [reply]
Ah I guess it was `\x`, here as C/C++ literal ... what a bad design idea. `\U` was not available back than; `\u` became standardized only in 99 it seems.

![image](https://github.com/user-attachments/assets/a045a3d5-8c9f-41a4-9f92-28c2ae6c567e)

**Reply by stamminator** [reply]
Thanks for the confirmation! I knew next to nothing about Unicode two weeks ago, so this has been a great crash course.

> which drove me crazy in like 1980 ;-)

My dad was in middle school in 1980 😳
